source/common/upstream/ring_hash_lb.cc

```diff
@@ -182,22 +185,31 @@ RingHashLoadBalancer::Ring::Ring(const Optional<envoy::api::v2::Cluster::RingHas
 
 void RingHashLoadBalancer::refresh() {
   // Note that we only compute global panic on host set refresh. Given that the runtime setting will
-  // rarely change, this is a reasonable compromise to avoid creating multiple rings when we only
-  // need to create one for LB.
-  RingConstSharedPtr new_ring;
-  bool new_global_panic;
-  const auto& host_set = chooseHostSet();
-  if (isGlobalPanic(host_set, runtime_)) {
-    new_ring = std::make_shared<Ring>(config_, host_set.hosts());
-    new_global_panic = true;
-  } else {
-    new_ring = std::make_shared<Ring>(config_, host_set.healthyHosts());
-    new_global_panic = false;
-  }
+  // rarely change, this is a reasonable compromise to avoid creating extra rings when we only
+  // need to create one per priority level.
+  for (auto& host_set : priority_set_.hostSetsPerPriority()) {
+    RingConstSharedPtr new_ring;
+    bool new_global_panic;
+    uint32_t priority = host_set->priority();
+    if (isGlobalPanic(*host_set, runtime_)) {
+      new_ring = std::make_shared<Ring>(config_, host_set->hosts());
+      new_global_panic = true;
+    } else {
+      new_ring = std::make_shared<Ring>(config_, host_set->healthyHosts());
+      new_global_panic = false;
+    }
 
-  std::unique_lock<std::shared_timed_mutex> lock(factory_->mutex_);
-  factory_->current_ring_ = new_ring;
-  factory_->global_panic_ = new_global_panic;
+    std::unique_lock<std::shared_timed_mutex> lock(factory_->mutex_);
```

Comment:
This happens on the main thread. You don't need the read lock here I don't think, you just need the write lock below when you swap it all in. You still need the read lock on the factory create call though.

Reply:
Above was a clear error, but this one confuses me.

You were locking factory_->mutex before changing factory->current_ring and factory->global_panic. Why then don't I need to snag the same lock when changing factory->per_priority_state?

Reply:
Sorry I read this too quickly. The issue here I think is that you might have inconsistent state since you release the lock several times during this code. I think the correct way to do this is:
1) Do the first loop to generate state, etc.
2) Acquire lock
3) Do 2nd loop to store state.